### PR TITLE
integration test for dynamic audit

### PR DESCRIPTION
**Pull request by pbarker**
Adds integration test for the dynamic audit feature

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by spiffxp** [comment]
/milestone v1.13
I'm adding this to the v1.13 milestone since it relates to https://github.com/kubernetes/features/issues/600 which is currently planned for v1.13

**Comment by spiffxp** [comment]
need rebase

**Review by jpbetz — CHANGES_REQUESTED** [review]
Seeing quite a few production code changes here which doesn't seem to match up with the PR title. Could we get an update to the production code changes made by this PR in the title/description?

**Comment by pbarker** [comment]
@CaoShuFeng PTAL at https://github.com/kubernetes/kubernetes/pull/69902/commits/8058975df2aa2b197f17d4b7c3b04c01ba2a5cad this effects some of the code you own here

**Comment by pbarker** [comment]
@jpbetz this is ready for a look 🙏

**Review by timothysc — COMMENTED** [review]
/approve 

but will defer on others to LGTM

File: test/integration/master/audit_dynamic_test.go
Comment (ncdc) [inline test/integration/master/audit_dynamic_test.go]: Please remove this blank line

File: test/integration/master/audit_dynamic_test.go
Comment (ncdc) [inline test/integration/master/audit_dynamic_test.go]: I'm concerned this potentially could flake in the future, depending on how overloaded the CI server is. It's probably ok to go with this for now, but please keep an eye on the *-bazel-verify jobs (which afaik run the integration tests) after this merges in case this starts to flake.

File: test/integration/master/audit_dynamic_test.go
Comment (ncdc) [inline test/integration/master/audit_dynamic_test.go]: Please add a comment that you're clearing event list 1 to start fresh for this test case

File: test/integration/master/audit_dynamic_test.go
Comment (ncdc) [inline test/integration/master/audit_dynamic_test.go]: Same possible flake concern as above; no changes required at this time

File: test/integration/master/audit_dynamic_test.go
Comment (ncdc) [inline test/integration/master/audit_dynamic_test.go]: You could use a mux for your test server and add a `/healthz` route that returns 200 for a `GET` request, instead of doing this

**Review by ncdc — CHANGES_REQUESTED** [review]

**Review by ncdc — APPROVED** [review]

File: test/integration/master/audit_dynamic_test.go
Comment (liggitt) [inline test/integration/master/audit_dynamic_test.go]: performing these immediately after registering the sink races with the dynamic backend observing the sink and registering the handler, right? need to do one of the following:
* wait for positive confirmation the sink is active (which we can't do because we don't expose that in the API)
* wait long enough after registering the sink that we would consider not observing it in that period a failure (5 seconds? 10 seconds?)
* retry the "configMapOperations && checkForEvents" operation in a loop until it passes or we exceed the time we consider acceptable for the sink to have become active

same comment applies other places a sink is added/removed/updated

Comment (ncdc) [reply test/integration/master/audit_dynamic_test.go]: This is similar to my comment about a potential flake risk. I think we could start by waiting for 5 seconds after the sink has been added/updated/deleted before performing operations. I talked with @pbarker about the possibility of doing this in a loop, but he wasn't sure if we could get to a clean state after running the loop multiple times. We can continue to discuss if needed.

Would you be ok with a sleep for starters?

File: test/integration/master/audit_dynamic_test.go
Comment (liggitt) [inline test/integration/master/audit_dynamic_test.go]: use PollImmediate with a much shorter interval (50ms?)... I'd expect the events to show up on the order of milliseconds

File: test/integration/master/audit_dynamic_test.go
Comment (liggitt) [reply test/integration/master/audit_dynamic_test.go]: > Would you be ok with a sleep for starters?

I could live with a 5 second timeout as long as you watch to make sure we're not hitting flakes, and look into a follow up that makes it more deterministic.

That said, I just looked at how long this new test is taking. These are the five run times for the test/integration/master package before this PR:
```
335.342s
323.410s
335.549s
336.179s
359.685s
```

and with this PR:
```
464.194s
449.109s
450.238s
467.853s
468.212s
```

This new test seems to be adding ~2 minutes to the test/integration/master package run time. That seems excessive (and long run times can make tests flake on timeouts as well). Is that runtime expected? Can we get it closer to ~20 seconds? The other 25 tests in the package complete much more quickly

Comment (pbarker) [reply test/integration/master/audit_dynamic_test.go]: ok this is updated, this was caused by waiting on the buffers to flush because we hadn't hit the max batch or wait time. Implemented @liggitt's req and moved the example test server into the framework, and modified the audit options so that the batch config could be hacked. Test takes about 13s now :)

File: test/integration/framework/BUILD
Comment (liggitt) [inline test/integration/framework/BUILD]: weird filename... can we call it `test_server.go`?

File: staging/src/k8s.io/apiserver/pkg/server/options/audit.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/audit.go]: set this in NewAuditOptions() as well?

File: test/integration/master/audit_dynamic_test.go
Comment (liggitt) [inline test/integration/master/audit_dynamic_test.go]: if the defaults are set in NewEditOptions, all this has to do is tweak the MaxBatchWait setting if `opts.Audit.DynamicOptions.BatchConfig` is not nil

File: test/integration/master/audit_dynamic_test.go
Comment (liggitt) [inline test/integration/master/audit_dynamic_test.go]: polls within polls are fun :)

this still seems potentially flaky. an event from the previous outer loop could sneak into the store, and if it did, checkForEvents would fail.

it might make more sense to do these checks in order:
1. is the sink actively receiving events (does the store have a non-zero number of items)
2. is the sink caught up (is the last event in the store the last thing we did)
3. did the sink see everything we expected? (the current checkForEvents helper)

we could do those checks like this after we register or change a sink:
1. create and delete a configmap with an incrementing name until the sink is non-empty
2. poll until the last item in the sink is the response event of the delete call for the last configmap we deleted
3. clear the sink store, do configMapOperations, call checkForEvents and poll until the sink observes the events

File: test/integration/master/audit_dynamic_test.go
Comment (liggitt) [inline test/integration/master/audit_dynamic_test.go]: I'm not used to seeing t.Run used serially like this... the tests seem fairly interdependent. if one fails won't the following ones fail too?

File: test/integration/framework/BUILD
Comment (pbarker) [reply test/integration/framework/BUILD]: not sure how that happened

File: test/integration/master/audit_dynamic_test.go
Comment (pbarker) [reply test/integration/master/audit_dynamic_test.go]: yeah they would, its more just a way to segment the individual pieces, we could remove it if it feels like its complicating things

File: test/integration/master/audit_dynamic_test.go
Comment (liggitt) [inline test/integration/master/audit_dynamic_test.go]: do we have already have tests verifying handoff from one sink to another doesn't duplicate or drop any events?

rather than looping and clearing the stores for the sink handoff here, I'd rather see something that:
1. rapidly produces unique events until sink2's store starts receiving events
2. takes the last event from sink1's store and the first from sink2's store, and verifies they are consecutive events
3. polls until sink2's store contains all the events we produced after it's first event
4. verifies no other further events were delivered to sink1

File: test/integration/master/audit_dynamic_test.go
Comment (pbarker) [reply test/integration/master/audit_dynamic_test.go]: I think I'll write a wrapper here that checks if the test passed and if not fail the parent

File: test/integration/master/audit_dynamic_test.go
Comment (pbarker) [reply test/integration/master/audit_dynamic_test.go]: should be atomic https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic.go#L231

File: test/integration/master/audit_dynamic_test.go
Comment (pbarker) [reply test/integration/master/audit_dynamic_test.go]: I thought about sleeping here but the method is used to check that the transition happens smoothly on the update. To guarantee that its better to ensure a continuous flow of events. It stops after 40 events, so I figured it was safe from any OOM scenario.

File: test/integration/master/audit_dynamic_test.go
Comment (tallclair) [inline test/integration/master/audit_dynamic_test.go]: This would be racy, except that there is only one thread that accesses the eventlist at a time. I'd prefer to just use a slice directly in that case.

File: test/integration/master/audit_dynamic_test.go
Comment (tallclair) [inline test/integration/master/audit_dynamic_test.go]: Picking back up the hotlooping thread - if there is an error in simpleOp for some reason, this will hotloop forever. I'm fine with a very short sleep - e.g. 1 millisecond.  Better yet - use a [rate limiter](https://godoc.org/github.com/kubernetes/kubernetes/staging/src/k8s.io/client-go/util/flowcontrol#NewTokenBucketRateLimiter) with sufficient burst to be effectively unthrottled in the happy path, but throttled in the worst case.

Comment (pbarker) [reply test/integration/master/audit_dynamic_test.go]: This client is currently set to `QPS:50 Burst:100`, do we think thats sufficient?

File: test/integration/master/audit_dynamic_test.go
Comment (pbarker) [reply test/integration/master/audit_dynamic_test.go]: nvm wrote a bad test to check this, adding `close`

File: test/integration/master/audit_dynamic_test.go
Comment (tallclair) [reply test/integration/master/audit_dynamic_test.go]: Got it. I guess it's fine as long as it's single-writer.

File: test/integration/master/audit_dynamic_test.go
Comment (tallclair) [reply test/integration/master/audit_dynamic_test.go]: Ah, that should be fine. Out of curiosity, where did you find those values?

**Comment by pbarker** [comment]
@liggitt @tallclair anything else on this one?

**Pull request merged**
